crates/stages/src/stages/headers.rs

```diff
@@ -54,56 +67,132 @@ where
+
+            if index > 0 && index % interval == 0 {
+                info!(target: "sync::stages::headers", progress = ((index as f64 / total_headers as f64) * 100.0).round(), "writing headers");
+            }
+
+            let (sealed_header, _) = SealedHeader::from_compact(&header_buf, header_buf.len());
+            let (header, header_hash) = sealed_header.split();
             if header.number == 0 {
                 continue
             }
+            last_header_number = header.number;
+
+            // Increase total difficulty
+            td += header.difficulty;
 
-            let header_hash = header.hash();
-            let header_number = header.number;
-            let header = header.unseal();
-            latest = Some(header.number);
+            // Header validation
+            self.consensus.validate_header_with_total_difficulty(&header, td).map_err(|error| {
+                StageError::Block {
+                    block: Box::new(header.clone().seal(header_hash)),
+                    error: BlockErrorKind::Validation(error),
+                }
+            })?;
 
-            // NOTE: HeaderNumbers are not sorted and can't be inserted with cursor.
-            tx.put::<tables::HeaderNumbers>(header_hash, header_number)?;
-            cursor_header.insert(header_number, header)?;
-            cursor_canonical.insert(header_number, header_hash)?;
+            // Append to HeaderTD
+            cursor_td.append(header.number, td.into())?;
+
+            // Append to CanonicalHeaders
+            cursor_canonical
+                .append(RawKey::<BlockNumber>::from_vec(number.clone()), header_hash.into())?;
+
+            // Append to Headers
+            cursor_header.append(RawKey::<BlockNumber>::from_vec(number), header.into())?;
+        }
+
+        info!(target: "sync::stages::headers", len = total_headers, "writing header hash index");
+
+        let mut cursor_header_numbers = tx.cursor_write::<RawTable<tables::HeaderNumbers>>()?;
+        let mut first_sync = false;
+
+        // If we only have the genesis block hash, then we are at first sync, and we can remove it,
+        // add it to the collector and use tx.append on all hashes.
+        if let Some((hash, block_number)) = cursor_header_numbers.last()? {
+            if block_number.value()? == 0 {
+                self.hash_collector.insert(hash.key()?, 0);
+                cursor_header_numbers.delete_current()?;
+                first_sync = true;
+            }
+        }
+
+        // Since ETL sorts all entries by hashes, we are either appending (first sync) or inserting
+        // in order (further syncs).
+        for (index, hash_to_number) in self.hash_collector.iter()?.enumerate() {
+            let (hash, number) = hash_to_number?;
+
+            if index > 0 && index % interval == 0 {
+                info!(target: "sync::stages::headers", progress = ((index as f64 / total_headers as f64) * 100.0).round(), "writing headers hash index");
```

Comment:
```suggestion
                info!(target: "sync::stages::headers", progress = ((index as f64 / total_headers as f64) * 100.0).round(), "Writing headers hash index");
```

+ same re percentage formatting